packages/react-router-dom/modules/__tests__/Link-test.js

```diff
@@ -63,6 +78,30 @@ describe('When a <Link> is clicked', () => {
 
   it('changes the location')
 
+  it('changes the location with "to" prop as a function', () => {
+    const node = document.createElement('div')
+
+    ReactDOM.render((
+      <MemoryRouter initialEntries={['/?world']}>
+        <div>
+          <Route path='/' exact render={({location}) => (
+            <Link to={location => ({...location, pathname: 'hello'})}>
+              link
+            </Link>
+          )} />
+          <Route path='/hello' render={({location}) => (
+            <div>{`Hello, ${location.search.substr(1)}`}</div>
+          )} />
+        </div>
+      </MemoryRouter>
+    ), node)
+
+    expect(node.textContent).not.toBe('Hello, world')
+    const a = node.getElementsByTagName('a')[0]
+    Simulate.click(a, { button: 0 })
+    expect(node.textContent).toBe('Hello, world')
```

Comment:
I would probably have the `<Link>`'s `to` function change the `search` string here so that we are verifying that `node.textContent` is changing. With the current test, we would have the same result if clicking the link did nothing.